streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -519,107 +520,169 @@ private long computeLatency() {
+    private long adjustRecordsProcessedBeforeCommit(final long prevRecordsProcessedBeforeCommit, final long totalProcessed,
+                                                    final long processLatency, final long commitTime) {
+        long recordsProcessedBeforeCommit = UNLIMITED_RECORDS;
+        // check if process latency larger than commit latency
+        // note that once we set recordsProcessedBeforeCommit, it will never be UNLIMITED_RECORDS again, so
+        // we will never process all records again. This might be an issue if the initial measurement
+        // was off due to a slow start.
+        if (processLatency > commitTime) {
+            // push down
+            recordsProcessedBeforeCommit = Math.max(1, (commitTime * totalProcessed) / processLatency);
+            log.debug("{} processing latency {} > commit time {} for {} records. Adjusting down recordsProcessedBeforeCommit={}",
+                logPrefix, processLatency, commitTime, totalProcessed, recordsProcessedBeforeCommit);
+        } else if (prevRecordsProcessedBeforeCommit != UNLIMITED_RECORDS && processLatency > 0) {
+            // push up
+            recordsProcessedBeforeCommit = Math.max(1, (commitTime * totalProcessed) / processLatency);
+            log.debug("{} processing latency {} > commit time {} for {} records. Adjusting up recordsProcessedBeforeCommit={}",
+                logPrefix, processLatency, commitTime, totalProcessed, recordsProcessedBeforeCommit);
+        }
 
-                        streamsMetrics.processTimeSensor.record(computeLatency(), timerStartedMs);
+        return recordsProcessedBeforeCommit;
+    }
 
-                        maybePunctuate(task);
+    /**
+     * Main event loop for polling, and processing records through topologies.
+     */
+    private void runLoop() {
+        long recordsProcessedBeforeCommit = UNLIMITED_RECORDS;
+        consumer.subscribe(sourceTopicPattern, rebalanceListener);
 
-                        if (task.commitNeeded())
-                            commitOne(task);
-                    }
+        while (stillRunning()) {
+            this.timerStartedMs = time.milliseconds();
 
-                } else {
-                    // even when no task is assigned, we must poll to get a task.
-                    requiresPoll = true;
+            // try to fetch some records if necessary
+            ConsumerRecords<byte[], byte[]> records = pollRequests(this.pollTimeMs);
+            if (records != null && !records.isEmpty() && !activeTasks.isEmpty()) {
+                streamsMetrics.pollTimeSensor.record(computeLatency(), timerStartedMs);
+                addRecordsToTasks(records);
+                final long totalProcessed = processAndPunctuate(activeTasks, recordsProcessedBeforeCommit);
+                if (totalProcessed > 0) {
+                    final long processLatency = computeLatency();
+                    streamsMetrics.processTimeSensor.record(processLatency / (double) totalProcessed,
```

Comment:
We still need it. For example, if `processAndPunctuate` is told to only process 5 records before committing (i.e., `recordsProcessedBeforeCommit` = 5), and we have 7 records in total, then `processAndPunctuate` would process 5, then commit, then process the 2 left and return 2. Then the above code in line 677 would record the latency of the last 2 records and attempt to commit if necessary.